rhine-terminal/src/FRP/Rhine/Terminal.hs

```diff
@@ -0,0 +1,165 @@
+{- | Wrapper to write @terminal@ applications in Rhine, using concurrency.
+-}
+
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE TypeFamilies #-}
+module FRP.Rhine.Terminal
+  ( TerminalConcT
+  , TerminalEventClock (..)
+  , launchTerminal
+  , flowTerminal
+  , terminalConcurrently
+  )
+  where
+
+-- base
+import Prelude hiding (putChar)
+import Control.Concurrent ()
+import Control.Concurrent.MVar ()
+import Data.IORef ()
+
+-- time
+import Data.Time.Clock ( getCurrentTime )
+
+-- transformers
+import Control.Monad.Trans.Class ( MonadTrans )
+import Control.Monad.Trans.Reader ( ReaderT(..), ask )
+
+-- terminal
+-- terminal
+import System.Terminal
+import System.Terminal.Internal ( LocalTerminal )
+
+-- rhine
+import FRP.Rhine
+    ( MonadIO(..),
+      constM,
+      liftTransS,
+      flow,
+      concurrently,
+      Arrow(first),
+      Clock(..),
+      HoistClock(..),
+      GetClockProxy,
+      In,
+      Out,
+      Schedule(..),
+      Rhine,
+      UTCTime )
+import Control.Monad.Catch (MonadMask, MonadCatch, MonadThrow)
+
+-- | Wraps the LocalTerminal value needed for communication with the @terminal@ backend.
+newtype TerminalConcT m a = TerminalConcT
+  { unTerminalConcT :: ReaderT LocalTerminal m a }
+  deriving (Functor, Applicative, Monad, MonadTrans, MonadIO, MonadThrow, MonadCatch, MonadMask)
+
+instance MonadInput (TerminalConcT IO) where
+  awaitWith = awaitWith
+
+instance MonadPrinter (TerminalConcT IO) where
+  putLn = putLn
+  putChar = putChar
+  putText = putText
+  getLineWidth = getLineWidth
```

Comment:
```suggestion
  awaitWith = awaitWith

instance MonadPrinter (TerminalConcT IO) where
  putLn = putLn
  putChar = putChar
  putText = putText
  getLineWidth = getLineWidth
```

I don't get these instances. They look like infinite loops to me. Don't you want to wrap/unwrap the newtype? Also you could try to derive `MonadPrinter` etc., does that not work with `GeneralizedNewtypeDeriving`?

Reply:
Ah we can't derive directly because `ReaderT` has no `MonadInput` instances and so on. So you need to write them by hand unfortunately.

Reply:
> They look like infinite loops to me. Don't you want to wrap/unwrap the newtype?

I thought I was doing something non-sensical here 😅